### PR TITLE
grub: Add support for the search command

### PR DESCRIPTION
**Pull request by rjoleary**
This adds support for the grub `search` command as defined by the grub
manual. This allows the kernel and initrd to be loaded from a file
system other from the boot drive. CentOS exercises this command and has
been added as a unit test.

The following options are included:

* `search --file`: Set root to the file system which contains the given
  file.
* `search --fs-label`: Set root to the file system with the given label.
* `search --fs-uuid`: Set root to the file system with the given UUID.

The aliases `search.file`, search.fs_label` and `search.fs_uuid` are
also supported.

Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

File: cmds/boot/boot/boot.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func main() {
 	if *verbose {
 		l = ulog.Log
 	}
-	images, mps, err := localboot.Localboot(l, blockDevs)
+	images, mountPool, err := localboot.Localboot(l, blockDevs)
```
Comment (rminnich) [inline cmds/boot/boot/boot.go:92]: in my heart of hearts I wish this were not part of this PR. It's unrelated to the subject.

File: pkg/boot/bootcmd/bootcmd.go [hunk 2]
```diff
@@ -17,10 +17,10 @@ import (
 // ShowMenuAndBoot handles common cleanup functions and flags that all boot
 // commands should support.
 //
-// mps are mounts to unmount before kexecing. noLoad prints the list of entries
+// mountPool is unmounted before kexecing. noLoad prints the list of entries
```
Comment (rminnich) [inline pkg/boot/bootcmd/bootcmd.go:20]: same comment

File: pkg/boot/boottest/json.go [hunk 3]
```diff
@@ -21,17 +21,45 @@ import (
...
+	"net/url"
 	"os"
+	"strings"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/u-root/u-root/pkg/boot"
 	"github.com/u-root/u-root/pkg/curl"
 )
 
+// makeFileRel converts file:/// schemes to a path relative to the current
```
Comment (rminnich) [inline pkg/boot/boottest/json.go:33]: Was this change absolutely required for the subject of the PR? I'm approving anyway but still ...

File: pkg/boot/boottest/json.go [hunk 4]
```diff
@@ -21,17 +21,45 @@ import (
...
+// working directory in order to make a hermatic comparison for the test. Note
```
Comment (rminnich) [inline pkg/boot/boottest/json.go:34]: hermetic.

File: pkg/boot/grub/config_test.go [hunk 5]
```diff
@@ -12,8 +12,47 @@ import (
...
 	"github.com/u-root/u-root/pkg/boot/boottest"
+	"github.com/u-root/u-root/pkg/mount"
+	"github.com/u-root/u-root/pkg/mount/block"
 )
 
+// fakeDevices returns a list of fake block devices and a pool of mount points.
+// The pool is pre-populated so that Mount is never called.
+func fakeDevices() (block.BlockDevices, *mount.Pool, error) {
+	// For some reason, Glob("testdata_new/*/") does not work here.
+	files, err := ioutil.ReadDir("testdata_new")
```
Comment (rminnich) [inline pkg/boot/grub/config_test.go:23]: likely because of the trailing / glob failed? Seems like it might be a bug?

**Review by rminnich — APPROVED** [review]
Wow. Nice.

File: pkg/boot/grub/config_test.go (hunk 5 above)
Comment (rjoleary) [reply pkg/boot/grub/config_test.go:23]: Not sure, but it's easy to reproduce:

```
echo 'package main; import "path/filepath"; import "fmt"; func main() { fmt.Println(filepath.Glob("/*/")); }' > main.go && go run main.go
```
vs

```
echo 'package main; import "path/filepath"; import "fmt"; func main() { fmt.Println(filepath.Glob("/*")); }' > main.go && go run main.go
```

The first prints nothing, while the second prints all files and directories. There's no way to Glob only directories.

File: cmds/boot/boot/boot.go (hunk 1 above)
Comment (rjoleary) [reply cmds/boot/boot/boot.go:92]: If we get some general agreement that the mount.Pool is a good approach for this, I can split it off into a separate PR to simplify the PRs (github isn't as nice as gerrit for patch series I'm afraid).

File: pkg/boot/boottest/json.go (hunk 3 above)
Comment (rjoleary) [reply pkg/boot/boottest/json.go:33]: Yes, the code takes a url.URL, converts it to a string, then parses it back to a url.URL. We were using a relative path in the url.URL previously which is non-standard, so Go adds an additional / in front which broke all the tests. Wikipedia says it best: "while "file:///foo.txt" is valid, "file://foo.txt" is not, although some interpreters manage to handle the latter".

File: pkg/boot/boottest/json.go [hunk 6]
```diff
@@ -21,17 +21,45 @@ import (
...
+	}
+	relU, err := url.Parse(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("error parsing %v: %v", u.String(), err)
+	}
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("error from os.GetWd(): %v", err)
+	}
+	relU.Path = strings.TrimPrefix(relU.Path, wd)
```
[the remark this answers is not in the document]
Comment (GanShun) [reply pkg/boot/boottest/json.go:49]: if so, add a comment

**Review by GanShun — DISMISSED** [review]

File: cmds/boot/boot/boot.go (hunk 1 above)
Comment (rjoleary) [reply cmds/boot/boot/boot.go:92]: I broke this off into #1860

File: pkg/boot/boottest/json.go (hunk 3 above)
Comment (rjoleary) [reply pkg/boot/boottest/json.go:33]: I split this out into another, dependent PR: https://github.com/u-root/u-root/pull/1865

**Comment by rjoleary** [comment]
Please take another look.

**Review by GanShun — APPROVED** [review]
One nit

File: pkg/boot/grub/grub.go [hunk 7]
```diff
@@ -298,14 +317,117 @@ func (c *parser) append(ctx context.Context, config string) error {
...
+				[]string{"search", map[string]string{
+					"search.file":     "--file",
+					"search.fs_label": "--fs-label",
+					"search.fs_uuid":  "--fs-uuid",
+				}[directive]},
+				kv[1:]...,
+			)
+			fallthrough
+		case "search":
+			fs := pflag.NewFlagSet("grub.search", pflag.ContinueOnError)
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/boot/grub/grub.go:334]: Done

**Review by GanShun — APPROVED** [review]
